File: blue-core/src/blue/noteProcessor/TimeWarpProcessor.java
package blue.noteProcessor;

import blue.BlueSystem;
import blue.plugin.NoteProcessorPlugin;
import blue.soundObject.Note;
import blue.soundObject.NoteList;
import blue.soundObject.NoteParseException;
import electric.xml.Element;

/**
 * 
 * Maps beats-to-seconds for time warping (accelerando, ritardando)
 * 
 * Based on code from OMDE (http://pythonsound.sourceforge.net/)
 * 
 * @author steven
 */

@NoteProcessorPlugin(displayName="TimeWarpProcessor", position = 130)
public class TimeWarpProcessor implements NoteProcessor, java.io.Serializable {

    private String timeWarpString = "0 60";

    public TimeWarpProcessor() {
    }

    /*
     * (non-Javadoc)
     * 
     * @see blue.noteProcessor.NoteProcessor#processNotes(blue.soundObject.NoteList)
     */
    @Override
    public void processNotes(NoteList in) throws NoteProcessorException {
        Note temp;
        TempoMapper tm = TempoMapper.createTempoMapper(this.timeWarpString);

        if (tm == null) {
            throw new NoteProcessorException(this, BlueSystem
                    .getString("noteProcessorException.tempoStringErr"));
        }

        float newStart, newEnd;

        for (int i = 0; i < in.size(); i++) {
            temp = in.get(i);
            try {
                newStart = tm.beatsToSeconds(temp.getStartTime());
                newEnd = tm.beatsToSeconds(temp.getStartTime()
                        + temp.getSubjectiveDuration());
            } catch (Exception ex) {
                throw new NoteProcessorException(this, BlueSystem
                        .getString("noteProcessorException.timeWarp"));
            }
            temp.setStartTime(newStart);
            if (newEnd - newStart < 0) {
                throw new NoteProcessorException(this, BlueSystem
                        .getString("noteProcessorException.timeWarp"));
            }
            temp.setSubjectiveDuration(newEnd - newStart);
        }
    }

    @Override
    public String toString() {
        return "[time warp]";
    }

    /**
     * @return
     */
    public String getTimeWarpString() {
        return timeWarpString;
    }

    /**
     * @param string
     */
    public void setTimeWarpString(String timeWarpString) {
        this.timeWarpString = timeWarpString;
    }

    // UNIT TEST

    public static void main(String[] args) {
        NoteList n = new NoteList();

        for (int i = 0; i < 10; i++) {
            try {
                n.add(Note.createNote("i1 " + i + " 1 3 4"));
            } catch (NoteParseException e) {
                // TODO Auto-generated catch block
                e.printStackTrace();
            }
        }

        System.out.println("before: \n\n" + n + "\n\n");

        TimeWarpProcessor twp = new TimeWarpProcessor();
        twp.setTimeWarpString("0 60 4 120");
        try {
            twp.processNotes(n);
        } catch (NoteProcessorException ex) {
            System.out.println("Exception: " + ex.getMessage());
        }

        System.out.println("after: \n\n" + n + "\n\n");
    }

    public static NoteProcessor loadFromXML(Element data) {
        TimeWarpProcessor twp = new TimeWarpProcessor();

        twp
                .setTimeWarpString(data.getElement("timeWarpString")
                        .getTextString());

        return twp;
    }

    /*
     * (non-Javadoc)
     * 
     * @see blue.noteProcessor.NoteProcessor#saveAsXML()
     */
    @Override
    public Element saveAsXML() {
        Element retVal = new Element("noteProcessor");
        retVal.setAttribute("type", this.getClass().getName());

        retVal.addElement("timeWarpString").setText(this.getTimeWarpString());

        return retVal;
    }

}

File: packages/blue-data/src/note-processors/time-warp-processor.ts
import { NoteProcessor } from './note-processor';
import { NoteProcessorException } from './note-processor-exception';
import { NoteList } from '../sound-objects/note-list';
import { Element } from '../serialization/xml-reader';

const JAVA_TYPE = 'blue.noteProcessor.TimeWarpProcessor';

export class TimeWarpProcessor extends NoteProcessor {
  private _timeWarpString = '0 60';

  constructor();
  constructor(src: TimeWarpProcessor);
  constructor(src?: TimeWarpProcessor) {
    super();
    if (src) {
      this._timeWarpString = src._timeWarpString;
    }
  }

  getTimeWarpString(): string {
    return this._timeWarpString;
  }
  setTimeWarpString(timeWarpString: string): void {
    this._timeWarpString = timeWarpString;
  }

  override process(notes: NoteList): NoteList {
    const tm = TempoMap.createTempoMap(this._timeWarpString);
    if (tm === null) {
      throw new NoteProcessorException('Error in tempo string', 0);
    }

    for (const note of notes) {
      let newStart: number;
      let newEnd: number;
      try {
        newStart = tm.beatsToSeconds(note.getStartTime());
        newEnd = tm.beatsToSeconds(note.getStartTime() + note.getSubjectiveDuration());
      } catch {
        throw new NoteProcessorException('Error in time warp', 0);
      }
      note.setStartTime(newStart);
      if (newEnd - newStart < 0) {
        throw new NoteProcessorException('Error in time warp', 0);
      }
      note.setSubjectiveDuration(newEnd - newStart);
    }
    return notes;
  }

  override getDisplayName(): string {
    return 'TimeWarpProcessor';
  }

  override deepCopy(): TimeWarpProcessor {
    return new TimeWarpProcessor(this);
  }

  saveAsXML(): Element {
    const elem = new Element('noteProcessor');
    elem.setAttribute('type', JAVA_TYPE);
    elem.addElement('timeWarpString').setText(this.getTimeWarpString());
    return elem;
  }

  static loadFromXML(data: Element): TimeWarpProcessor {
    const proc = new TimeWarpProcessor();
    const tws = data.getTextString('timeWarpString');
    if (tws !== null) proc._timeWarpString = tws;
    return proc;
  }
}

class BeatTempoPair {
  beat = 0;
  tempo = 0;
}

class TempoMap {
  private _pairs: BeatTempoPair[];

  private constructor(pairs: BeatTempoPair[]) {
    this._pairs = pairs;
  }

  static createTempoMap(tempoString: string): TempoMap | null {
    const tokens = tempoString.trim().split(/\s+/);
    if (tokens.length % 2 !== 0) return null;

    const pairs: BeatTempoPair[] = [];
    for (let i = 0; i < tokens.length; i += 2) {
      const pair = new BeatTempoPair();
      pair.beat = parseFloat(tokens[i]);
      pair.tempo = parseFloat(tokens[i + 1]);
      if (isNaN(pair.beat) || isNaN(pair.tempo) || pair.beat < 0) {
        return null;
      }
      pairs.push(pair);
    }
    if (pairs.length === 0) return null;
    return new TempoMap(pairs);
  }

  beatsToSeconds(beat: number): number {
    if (this._pairs.length === 0) return 0;

    let seconds = 0;

    for (let i = 0; i < this._pairs.length - 1; i++) {
      const current = this._pairs[i];
      const next = this._pairs[i + 1];

      if (beat <= next.beat) {
        const bpm = (current.tempo + next.tempo) / 2;
        const secondsPerBeat = 60 / bpm;
        return seconds + (beat - current.beat) * secondsPerBeat;
      }

      const bpm = (current.tempo + next.tempo) / 2;
      const secondsPerBeat = 60 / bpm;
      seconds += (next.beat - current.beat) * secondsPerBeat;
    }

    const last = this._pairs[this._pairs.length - 1];
    const secondsPerBeat = 60 / last.tempo;
    return seconds + (beat - last.beat) * secondsPerBeat;
  }
}
